### PR TITLE
Add critical section API wrappers

### PR DESCRIPTION
**Pull request by ngoldbaum**
ping @bschoenmaeckers for your other PR.

Adds `Default` impls for `ffi::PyCriticalSection` and `ffi::PyCriticalSection2`. These are needed to actually use these structs, since the fields are private.

I'm not sure how to write a test to check that critical sections are closed when a panic happens. I think I'd need to instrument the RAII guard for that to work, and that seems wrong to me.

I'd also appreciate suggestions for an example to go in the docstring.

### CODE REVIEW

**Comment by bschoenmaeckers** [comment]
Test currently fails on [emscripten](https://github.com/PyO3/pyo3/actions/runs/11121369599/job/30900361369#logs).

File: src/sync.rs [hunk 1]
```diff
@@ -330,11 +330,67 @@ impl Interned {
...
+                    crate::ffi::PyCriticalSection_End(&mut self.0);
+                }
+            }
+        }
+
+        let _guard = unsafe {
+            let mut section = std::mem::zeroed();
+            crate::ffi::PyCriticalSection_Begin(&mut section, object.as_ptr());
+            Guard(section)
+        };
```
Comment (mejrs) [inline src/sync.rs:374]: You should not be moving this
```suggestion
        let mut guard = Guard(unsafe {std::mem::zeroed()});
        unsafe { crate::ffi::PyCriticalSection_Begin(&mut guard.section, object.as_ptr()) };
```

Comment (bschoenmaeckers) [reply src/sync.rs:374]: Should we add `PhantomPinned` to `PyCriticalSection`?

**Comment by ngoldbaum** [comment]
Added the ci-build-full label to see if there are any other corner cases in the full tests.

File: src/sync.rs [hunk 2]
```diff
@@ -330,11 +330,58 @@ impl Interned {
     }
 }
 
+/// Executes a closure with a Python critical section held on an object.
+///
+/// Acquires the per-object lock for the object `op` that is held
+/// until the closure `f` is finished.
+///
+/// This is structurally equivalent to the use of the paired
+/// Py_BEGIN_CRITICAL_SECTION and Py_END_CRITICAL_SECTION macros.
```
Comment (davidhewitt) [inline src/sync.rs:339]:
```suggestion
/// Py_BEGIN_CRITICAL_SECTION and Py_END_CRITICAL_SECTION C-API macros.
```

Maybe let's also link to https://docs.python.org/3.13/c-api/init.html#python-critical-section-api somewhere in this doc?

File: pyo3-ffi/src/cpython/critical_section.rs [hunk 3]
```diff
@@ -9,13 +9,47 @@ pub struct PyCriticalSection {
...
+        }
+    }
+}
+
+#[cfg(Py_GIL_DISABLED)]
+impl Default for PyCriticalSection {
+    fn default() -> Self {
+        PyCriticalSection::new()
+    }
+}
```
Comment (davidhewitt) [inline pyo3-ffi/src/cpython/critical_section.rs:27]: I guess all these new / default values weren't needed, as you used `mem::zeroed`?

**Review by davidhewitt — APPROVED** [review]
Overall looks great, thanks!

File: pyo3-ffi/src/cpython/critical_section.rs (hunk 3 above)
Comment (davidhewitt) [reply pyo3-ffi/src/cpython/critical_section.rs]: I pushed a commit to revert these for now.

File: src/sync.rs (hunk 1 above)
Comment (mejrs) [reply src/sync.rs]: > Should we add `PhantomPinned` to `PyCriticalSection`?

It doesn't really matter in this case.

Implementing a type that "needs pinning" requires unsafe code, and writing code that needs pinned types also requires unsafe code. Pin is just a way to let safe code carry that promise between two pieces of unsafe code. The latter doesn't happen here.
